sound/soc/sof/intel/hda.h

```diff
@@ -392,6 +392,14 @@ struct sof_intel_dsp_bdl {
 #define SOF_HDA_PLAYBACK		0
 #define SOF_HDA_CAPTURE			1
 
+/*
+ * Time in ms for opportunistic D0I3 entry delay.
+ * This has been deliberately chosen to be longer than the runtime suspend
+ * delay in order to make sure that the delayed work to enter D0I3 gets
+ * executed only if the SOF runtime suspend does not kick in.
```

Comment:
I am not fully sold on this. Imagine a case where we'd want to do low-power playback with a 1s buffer. We'd want to enter D0i3 very quickly and only transition back to D0 when we need to refill the buffer. With such a long timeout, we would never enter D0i3.

Reply:
@plbossart I think theoretically choosing a smaller timeout for the D0I3 work in S0 is just fine. D0I3/S0 and runtime suspend are mutually exclusive anyway, so there's no need for a relationship between the 2 timeouts. But I suppose the question is what happens during stress tests where we open/close streams in 1-2s intervals. We would be entering D0I3 only to exit when the subsequent stream is opened. Is that ok?

Reply:
@plbossart what is your opinion here? Should I change it to a smaller value ?

Reply:
honestly not sure. The longer values are interesting in that you can pretty much avoid race conditions. If the values are smaller, then the 'last_busy' case isn't always clear and we might get in trouble? I am especially concerned about cases with child devices (SoundWire, etc), where control of the pm_runtime status isn't exactly straightforward (as shown with race conditions where the parent didn't suspend but the child did, etc).

Reply:
Could we clever here and tie this in with NO_IRQ and large stream buffer sizes. i.e. I would expect a low power application to use a large buffer with no IRQ mode enabled.  e.g. playback stream with NO_IRQ enabled then enter D0I3 immediately after user updates buffer pointer OR playback stream using file IO, then enter D0I3 immediately if period size > X. This would ensure we are not thrashing PM transitions.

We really need to stay in D0I3 for as long as possible, so fast entry is needed when conditions are met.  This also helps to maximise CPU wait state depth and save further power.

Reply:
Thinking about this a bit more, we'd never see a conflict between pm_runtime suspend and D0ix while doing low-power playback. the refcount guarantees that pm_runtime will never be used, so we can use a very low threshold to allow for D0ix transitions.

Even with hotwording, the device is active so runtime_pm is not in scope.

So I guess the main issue is what happens in transients where we go to D0ix but then pm_runtime suspend. The platform-specific drivers could take care of going back to D0, no?

Reply:
@plbossart yes that is correct. runtime suspend and D0i3 are mutually exclusive. And we always resume to D0 before entering runtime suspend. The IPC that would be sent to close the D0i3-compat stream would resume the DSP to D0.
So we could just set a small enough timeout for the D0I3 work to be scheduled to ensure it meets the requirements  for low-power playback. But what should this timeout be in this case?

Reply:
@ranj063 @plbossart this small transition timeout will probably be platform specific (given the differences in doorbell and mailbox IPC mechanism and DSP capabilities between platforms).